pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java

```diff
@@ -191,6 +192,66 @@ public static boolean shouldUseVarLengthDictionary(FieldSpec.DataType columnStor
     return false;
   }
 
+  /**
+   * This function evaluates whether to override dictionary (i.e use noDictionary)
+   * for a column even when its explicitly configured. This evaluation is for both dimension and metric
+   * column types.
+   * @return
+   */
+  public static boolean ignoreDictionaryOverride(boolean optimizeDictionary,
+      boolean optimizeDictionaryForMetrics, double noDictionarySizeRatioThreshold,
+      FieldSpec fieldSpec, FieldIndexConfigs fieldIndexConfigs, int colCardinality,
+      int totalNumberOfEntries) {
+    if (optimizeDictionary) {
+      // Do not create dictionaries for json or text index columns as they are high-cardinality values almost always
+      if ((fieldIndexConfigs.getConfig(StandardIndexes.json()).isEnabled() || fieldIndexConfigs.getConfig(
+          StandardIndexes.text()).isEnabled())) {
+        return false;
+      }
+      // Do not create dictionary if index size with dictionary is going to be larger than index size without dictionary
+      // This is done to reduce the cost of dictionary for high cardinality columns
+      // Off by default and needs optimizeDictionary to be set to true
+      if (fieldSpec.isSingleValueField() && fieldSpec.getDataType().isFixedWidth()) {
+        // if you can safely enable dictionary, you can ignore overrides
+        return canSafelyEnableDictionary(colCardinality, totalNumberOfEntries,
+            noDictionarySizeRatioThreshold, fieldSpec);
+      }
+    }
+
+    if (optimizeDictionaryForMetrics && !optimizeDictionary) {
+      if (fieldSpec.isSingleValueField() && fieldSpec.getDataType().isFixedWidth() && fieldSpec.getFieldType()
+          == FieldSpec.FieldType.METRIC) {
+        // if you can safely enable dictionary, you can ignore overrides
+        return canSafelyEnableDictionary(colCardinality, totalNumberOfEntries,
+            noDictionarySizeRatioThreshold, fieldSpec);
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Given the column cardinality, totalNumberOfEntries, this function checks if the savings ratio
+   * is larger than the configured threshold (noDictionarySizeRatioThreshold). If savings ratio is
+   * smaller than the threshold, we want to override to noDictionary.
+   * @return
+   */
+  private static boolean canSafelyEnableDictionary(int colCardinality, int totalNumberOfEntries,
```

Comment:
nit: can we use the original function name here or use something like: `canSafelyCreateDictionaryWithinThreshold` to indicate something about the threshold checks.